lib/OpenQA/WebAPI/Controller/File.pm

```diff
@@ -108,17 +109,12 @@ sub test_asset {
     }
 
     # find the asset path
-    my $path = locate_asset($asset{type}, $asset{name}, 0);
-    # map to URL
-    my $urlname = $path =~ m,/fixed/, ? 'download_fixed_asset' : 'download_asset';
-    $path = $self->url_for($urlname, assettype => $asset{type}, assetname => $asset{name});
-    if ($self->param('subpath')) {
-        $path .= '/' . $self->param('subpath');
-        # better safe than sorry. Mojo seems to canonicalize the
-        # urls for us already so this is actually not needed
-        return $self->render_exception('invalid character in path') if ($path =~ /\/\.\./ || $path =~ /\.\.\//);
-    }
+    my $path = locate_asset($asset{type}, $asset{name}, relative => 1);
+    $path = catfile($path, $self->param('subpath')) if $self->param('subpath');
+    return $self->render(text => 'invalid character in path', status => 400) if ($path =~ /\/\.\./ || $path =~ /\.\.\//);
```

Comment:
It's a nitpick, and to follow my own rules should maybe be a separate commit, but this message is dumb. It's not about an 'invalid character' at all. The error message should reflect what it's *actually* disallowing.